pkg/metadata/manager.go

```diff
@@ -0,0 +1,404 @@
+package metadata
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+	"sync"
+	"sync/atomic"
+
+	"github.com/cilium/cilium/pkg/cgroups"
+	"github.com/cilium/cilium/pkg/defaults"
+	v1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	nodetypes "github.com/cilium/cilium/pkg/node/types"
+	"github.com/cilium/cilium/pkg/option"
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "metadata")
+	// example cgroup path in kubernetes environments
+	// /run/cilium/cgroupv2/kubepods/burstable/pod1858680e-b044-4fd5-9dd4-f137e30e2180/e275d1a37782ab30008aa3ae6666cccefe53b3a14a2ab5a8dc459939107c8c0
+	defaultCgroupBasePath = cgroups.GetCgroupRoot() + "/kubepods"
+	// example cgroup path in nested kubernetes environments like kind
+	// /run/cilium/cgroupv2/kubelet/kubepods/pod4841248b-fc2f-41f4-9981-a685bf840ab5/d8f227cc24940cfdce8d8e601f3b92242ac9661b0e83f0ea57fdea1cb6bc93ec
+	defaultNestedCgroupBasePath = cgroups.GetCgroupRoot() + "/kubelet" + "/kubepods"
+	cgroupBasePaths             = []string{defaultCgroupBasePath, defaultNestedCgroupBasePath}
+)
+
+// Manager maintains Kubernetes and low-level metadata (cgroup path and
+// cgroup id) for local pods and their containers. In order to do that, it defines
+// and implements callback functions that are called on Kubernetes pod watcher events.
+// It also exposes APIs to read the saved metadata.
+//
+// During initialization, the manager checks for a valid base cgroup path.
+// In case of environments using non-default cgroup base paths, manager will
+// fail to get a valid cgroup base path, and ignore all the subsequent pod events.
+type Manager struct {
```

Comment:
If I follow correctly, the internals of this structure *must* be accessed only by the main goroutine inside the manager, and that's how we keep it lockless (basically by implementing message passing via podEvents for any/all interactions). It's probably worth mentioning this in a comment above the structure here as a hint to future reviewers that other goroutines must not start interacting with these fields.